gmso/external/convert_foyer.py

```diff
@@ -0,0 +1,582 @@
+    """Convert a gmso XML to a foyer XML
+
+    Parameters
+    ----------
+    gmso_xml : XML
+        An XML file in the GMSO format 
+   
+    Returns
+    -------
+    foyer_xml : XML
+        An XML file in the foyer format
+    """
+
+    gmso_xml_tree = etree.parse(gmso_xml)
+    f_kwargs = {
+            'ffmetadata': [],
+            'atom_type_units': [],
+            'atom_types': [],
+            'non_bonded_forces': [],
+            'bond_type_units': [],
+            'harmonic_bond_types': [],
+            'angle_type_units': [],
+            'harmonic_angle_types': [],
+            'urey_bradley_angle_types': [],
+            'dihedral_units': [],
+            'periodic_torsion_dihedral_types': [],
+            'rb_torsion_dihedral_types': [],
+            }
+
+    # Try to load in FF Metadata
+    ff_metadata = gmso_xml_tree.findall('FFMetaData')[0]
+    ff_units = ff_metadata.findall('Units')[0]
+    units = ['energy', 'mass', 'charge', 'distance']
+    energy_dict = {typ: ff_units.get(typ) for typ in units}
+    f_kwargs['ffmetadata'] = energy_dict
+
+    # Try to load in AtomType section
+    # Load in AtomTypes section otherwise
+    atom_types_el = gmso_xml_tree.findall('AtomTypes')
+    for atom_types in atom_types_el:
+        for atom_type in atom_types.getiterator('AtomType'):
+            f_kwargs['atom_types'].append(atom_type)
+
+    for units in atom_types_el:
+        param_dict = {}
+        for unit in units.getiterator('ParametersUnitDef'):
+            param_dict[unit.get('parameter')] = unit.get('unit')
+
+    f_kwargs['atom_type_units'] = param_dict
+
+    harmonic_bond_force_el = gmso_xml_tree.findall('BondTypes')
+    for hbf in harmonic_bond_force_el:
+        for bond_type in hbf.getiterator('BondType'):
+            f_kwargs['harmonic_bond_types'].append(bond_type)
+
+    for units in harmonic_bond_force_el:
+        param_dict = {}
+        for unit in units.getiterator('ParametersUnitDef'):
+            param_dict[unit.get('parameter')] = unit.get('unit')
+
+    f_kwargs['bond_type_units'] = param_dict
+
+    harmonic_angle_force_el = gmso_xml_tree.findall('AngleTypes')
+    for haf in harmonic_angle_force_el:
+        for angle_type in haf.getiterator('AngleType'):
+            f_kwargs['harmonic_angle_types'].append(angle_type)
+
+    param_dict = {}
+    for units in harmonic_angle_force_el:
+        for unit in units.getiterator('ParametersUnitDef'):
+            param_dict[unit.get('parameter')] = unit.get('unit')
+
+    f_kwargs['angle_type_units'] = param_dict
+
+    urey_bradley_angle_el = gmso_xml_tree.findall('AmoebaUreyBradleyForce')
+    for ubf in urey_bradley_angle_el:
+        for angle_type in ubf.getiterator('UreyBradley'):
+            f_kwargs['urey_bradley_angle_types'].append(angle_type)
+
+    dihedral_torsion_force_el = gmso_xml_tree.findall('DihedralTypes')
+    for tf in dihedral_torsion_force_el:
+        if 'k * (1 + cos(n * phi - delta))' in tf.get('expression'):
```

Comment:
I dont know if this would work if the same expression was there, but shuffled around a bit. We might need to use sympify here?